src/libraries/Common/src/System/Security/Cryptography/ECDiffieHellmanOpenSsl.Derive.cs

```diff
@@ -123,79 +131,64 @@ public override byte[] DeriveRawSecretAgreement(ECDiffieHellmanPublicKey otherPa
                 {
-                    ourKey = _key.UpRefKeyHandle();
                     theirKey = otherKey.DuplicateKeyHandle();
                 }
                 else if (otherIsNamed)
                 {
-                    ourKey = _key.UpRefKeyHandle();
-
                     using (ECOpenSsl tmp = new ECOpenSsl(otherKey.ExportExplicitParameters()))
                     {
-                        theirKey = tmp.UpRefKeyHandle();
+                        theirKey = tmp.CreateEvpPKeyHandle();
                     }
                 }
                 else
                 {
-                    using (ECOpenSsl tmp = new ECOpenSsl(thisKeyExplicit))
+                    try
                     {
-                        ourKey = tmp.UpRefKeyHandle();
+                        // This is generally not expected to fail except:
+                        // - when key can't be accessed but is available (i.e. TPM)
+                        // - private key is actually missing
+                        using (ECOpenSsl tmp = new ECOpenSsl(ExportExplicitParameters(true)))
+                        {
+                            ourKey = tmp.CreateEvpPKeyHandle();
+                            disposeOurKey = true;
+                        }
                     }
-
-                    theirKey = otherKey.DuplicateKeyHandle();
-                }
-
-                using (SafeEvpPKeyCtxHandle ctx = Interop.Crypto.EvpPKeyCtxCreate(ourKey, theirKey, out uint secretLengthU))
-                {
-                    if (ctx == null || ctx.IsInvalid || secretLengthU == 0 || secretLengthU > int.MaxValue)
+                    catch (CryptographicException)
                     {
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                        // In both cases of failure we'll report lack of private key
+                        throw new CryptographicException(SR.Cryptography_CSP_NoPrivateKey);
                     }
 
-                    secretLength = (int)secretLengthU;
-
-                    // Indicate that secret can hold stackallocs from nested scopes
-                    scoped Span<byte> secret;
-
-                    // Arbitrary limit. But it covers secp521r1, which is the biggest common case.
-                    const int StackAllocMax = 66;
-
-                    if (secretLength > StackAllocMax)
-                    {
-                        rented = CryptoPool.Rent(secretLength);
-                        secret = new Span<byte>(rented, 0, secretLength);
-                    }
-                    else
-                    {
-                        secret = stackalloc byte[secretLength];
-                    }
+                    theirKey = otherKey.DuplicateKeyHandle();
+                }
 
-                    Interop.Crypto.EvpPKeyDeriveSecretAgreement(ctx, secret);
+                int written = Interop.Crypto.EvpPKeyDeriveSecretAgreement(ourKey, theirKey, secret);
```

Comment:
I know above you say 128 will always be enough, but do we need an assert to help validate that?

Reply:
EvpPKeyDeriveSecretAgreement will reject if it's too small

Reply:
Meaning it will throw? If so, ok.

Reply:
yes, EvpPKeyDeriveSecretAgreement is a tiny wrapper on top of a `CryptoNative_EvpPKeyDeriveSecretAgreement` which ends up calling EVP_PKEY_derive and will throw CryptographicException when that returns 0 (and I validated in C EVP_PKEY_derive actually does detect smaller buffer)